feat(source): Add 'kn source list'

navidshaikh opened the pull request:
/lint
Fixes #480

## Proposed Changes
- Add 'kn source list' listing the available sources COs
- Use dynamic client to find available source objects
- Add --type flag to filter source list based on source type
- Printed a source with machine readable format:
   - If source objects are of one type: `apiVersion` and `kind` is set to respective source type's object
   - If source objects are of different types in list: `apiVersion` is set to `""` and `kind` is to `List`.
   - If source objects are of different types but only one `--type` filter is given to subset source objects, the machine readable format sets `apiVersion` and `List` to respective source's type.

#### Default output:
```
kn source list
NAME   TYPE              RESOURCE                                        SINK         READY
a1     ApiServerSource   apiserversources.sources.eventing.knative.dev   svc:eshow2   True
c1     CronJobSource     cronjobsources.sources.eventing.knative.dev     svc:eshow3   True
p1     PingSource        pingsources.sources.knative.dev                 svc:eshow1   True
```

#### Using (case insensitive) type filters:

1. Single source type
```
➜  client git:(pr/source-list) kn source list --type ApiServerSource
NAME   TYPE              RESOURCE                                        SINK         READY
a1     ApiServerSource   apiserversources.sources.eventing.knative.dev   svc:eshow2   True
```

2. Multiple source types
```
➜  client git:(pr/source-list) kn source list --type ApiServerSource --type pingsource
NAME   TYPE              RESOURCE                                        SINK         READY
a1     ApiServerSource   apiserversources.sources.eventing.knative.dev   svc:eshow2   True
p1     PingSource        pingsources.sources.knative.dev                 svc:eshow1   True
```

#### See `kind` and `apiVersion` of `List` for machine readable format:

1. For multiple types of source objects
```
➜  client git:(pr/source-list) kn source list -oyaml| grep kind | tail -1
kind: List
```

```
➜  client git:(pr/source-list) kn source list -oyaml| grep apiVersion | head -1
apiVersion: ""
```

2. For single source type objects (or sub-set a single type using `--type` filter):
```
➜  client git:(pr/source-list) kn source list --type apiserversource -oyaml | grep kind | tail -1 
kind: ApiServerSourceList
```

```
➜  client git:(pr/source-list) kn source list --type apiserversource -oyaml | grep apiVersion | head -1
apiVersion: sources.eventing.knative.dev/v1alpha1
```

navidshaikh commented:
Adding another example of `kn source list` with another column `READY`.

navidshaikh commented:
With 3 additional columns: `CONDITIONS, READY, REASON`

```
kn source list
NAME              TYPE              RESOURCE                                        SINK        CONDITIONS   READY   REASON
k8sevents         ApiServerSource   apiserversources.sources.eventing.knative.dev   svc:mysvc   6 OK / 6     True    
k8sevents2        ApiServerSource   apiserversources.sources.eventing.knative.dev   svc:mysvc   2 OK / 5     False    : User system:serviceaccount:default:doesnt-exist cannot get, list, watch resource "events" in API group ""
my-cron-trigger   CronJobSource     cronjobsources.sources.eventing.knative.dev     svc:mysvc   7 OK / 7     True    
```

navidshaikh commented:
With only `READY` column 

```
kn source list
NAME              TYPE              RESOURCE                                        SINK        READY
k8sevents         ApiServerSource   apiserversources.sources.eventing.knative.dev   svc:mysvc   True
k8sevents2        ApiServerSource   apiserversources.sources.eventing.knative.dev   svc:mysvc   False
my-cron-trigger   CronJobSource     cronjobsources.sources.eventing.knative.dev     svc:mysvc   True
```

IMO `READY` column is enough else the columns don't fit in screen, `CONDITIONS` and `READY` can be seen from `describe` instead.

navidshaikh commented:
Added `--type` flag to filter source list based on source type(s). Updated PR description with examples.

navidshaikh commented:
/lint

navidshaikh commented:
This is ready for review.
/cc @rhuss @daisy-ycguo

navidshaikh commented:
/test pull-knative-client-integration-tests

```
no endpoints available for service "webhook"
```

navidshaikh commented:
/retest

On Tue, 18 Feb 2020 at 9:06 PM, Knative Prow Robot <notifications@github.com>
wrote:

> @navidshaikh <https://github.com/navidshaikh>: The following test *failed*,
> say /retest to rerun all failed tests:
> Test name Commit Details Rerun command
> pull-knative-client-integration-tests 4d74e8d
> <https://github.com/knative/client/commit/4d74e8ddf94b3e9df56cf9c9f9ddaaa5835ae5e2>
> link
> <https://prow.knative.dev/view/gcs/knative-prow/pr-logs/pull/knative_client/666/pull-knative-client-integration-tests/1229785508830253057> /test
> pull-knative-client-integration-tests
>
> Full PR test history
> <https://gubernator.knative.dev/pr/knative_client/666>. Your PR dashboard
> <https://gubernator.knative.dev/pr/navidshaikh>.
>
> Instructions for interacting with me using PR comments are available here
> <https://git.k8s.io/community/contributors/guide/pull-requests.md>. If
> you have questions or suggestions related to my behavior, please file an
> issue against the kubernetes/test-infra
> <https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:>
> repository. I understand the commands that are listed here
> <https://go.k8s.io/bot-commands>.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/knative/client/pull/666?email_source=notifications&email_token=AATE22ACJQRCZ4GBEPFXFN3RDP577A5CNFSM4KVCVIL2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMCOBWQ#issuecomment-587522266>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AATE22BBJSL7JXWIU3HFEXLRDP577ANCNFSM4KVCVILQ>
> .
>
-- 
Navid Shaikh
Composed on handheld device.

navidshaikh commented:
/retest

```
smoketests: timeout: service 'hello' not ready after 600 seconds
```

rhuss commented on pkg/dynamic/client.go:
```diff
@@ -94,3 +111,86 @@ func (c *knDynamicClient) ListSourcesTypes() (*unstructured.UnstructuredList, er
+		if len(sList.Items) > 0 {
+			sourceList.Items = append(sourceList.Items, sList.Items...)
+		}
+		sourceList.SetGroupVersionKind(sList.GetObjectKind().GroupVersionKind())
```
this looks wrong to me. The `sourceList` contains all elements from any type, so it looks for me that the last type just set the type of the sourceList (which is wrong as it can contain objects with different types)

Instead I would use for `sourceList` just a `[]*Unstructured` intstead of the list object (as we don't have a homogenous type)

navidshaikh replied:
Yes, it sets the gvk for the last item processed. 

> Instead I would use for sourceList just a []*Unstructured intstead of the list object

Sorry, I didnt get this. 
Do you mean to say, don't set the GVK for sourceList?
(we needed GVK for machine readable flags for printers)

rhuss replied:
I mean, don't use a `UnstructuredList` for the return value (except there is a specific reason for this, e.g. for the tabbed wrier), but a `[]*Unstructured` (so, only the `.items`), as the GVK for this list doesn't make sense for a potentially heterogenous list with different GVKs

navidshaikh replied:
json/yaml/name printer requires the non `Empty` GVK for an object (List in this case), to print machinereadable output.

With following change: (i.e. setting `{Group: "", Version: "v1alpha1", Kind: "List"}` once for this list)
```
diff --git a/pkg/dynamic/client.go b/pkg/dynamic/client.go
index 1fc6e91a..71511343 100644
--- a/pkg/dynamic/client.go
+++ b/pkg/dynamic/client.go
@@ -151,8 +151,9 @@ func (c *knDynamicClient) ListSources(f *SourceListFilters) (*unstructured.Unstr
                if len(sList.Items) > 0 {
                        sourceList.Items = append(sourceList.Items, sList.Items...)
                }
-               sourceList.SetGroupVersionKind(sList.GetObjectKind().GroupVersionKind())
+
        }
+       sourceList.SetGroupVersionKind(schema.GroupVersionKind{Group: "", Version: "v1alpha1", Kind: "List"})
        return &sourceList, nil
 }
```
produce:

```
apiVersion: v1alpha1
items:
- apiVersion: sources.eventing.knative.dev/v1alpha1
  kind: ApiServerSource
  metadata:
    annotations:
      sources.eventing.knative.dev/creator: kube:admin
      sources.eventing.knative.dev/lastModifier: kube:admin
    creationTimestamp: "2020-02-25T16:58:36Z"
    generation: 1
    name: k8sevents
    namespace: default
    resourceVersion: "363874"
    selfLink: /apis/sources.eventing.knative.dev/v1alpha1/namespaces/default/apiserversources/k8sevents
    uid: 0e0bd443-5965-47e5-8ee5-76382208f5dc
  spec:
    mode: Ref
    resources:
    - apiVersion: v1
      controller: false
      controllerSelector:
        apiVersion: ""
        kind: ""
        name: ""
        uid: ""
      kind: Event
      labelSelector: {}
    serviceAccountName: events-sa
    sink:
      ref:
        apiVersion: serving.knative.dev/v1
        kind: Service
        name: eshow1
        namespace: default
  status:
    conditions:
    - lastTransitionTime: "2020-02-25T16:59:20Z"
      status: "True"
      type: Deployed
    - lastTransitionTime: "2020-02-25T16:59:34Z"
      message: apiserversources.sources.eventing.knative.dev are deprecated and will
        be removed in the future. Use apiserversources.sources.knative.dev instead.
      reason: ApiServerSourceDeprecated
      severity: Warning
      status: "True"
      type: Deprecated
    - lastTransitionTime: "2020-02-25T16:58:53Z"
      severity: Info
      status: "True"
      type: EventTypesProvided
    - lastTransitionTime: "2020-02-25T16:59:20Z"
      status: "True"
      type: Ready
    - lastTransitionTime: "2020-02-25T16:58:53Z"
      status: "True"
      type: SinkProvided
    - lastTransitionTime: "2020-02-25T16:58:53Z"
      status: "True"
      type: SufficientPermissions
    observedGeneration: 1
    sinkUri: http://eshow1.default.svc.cluster.local
kind: List
```
which is kind of similar to what `kubectl get sources -oyaml` produces (it also sets `Group` to "" and `apiVersion` to `v1`)
```
✗ kubectl get sources -oyaml  
apiVersion: v1
items:
- apiVersion: sources.eventing.knative.dev/v1alpha1
  kind: ApiServerSource
  metadata:
    annotations:
      sources.eventing.knative.dev/creator: kube:admin
      sources.eventing.knative.dev/lastModifier: kube:admin
    creationTimestamp: "2020-02-25T16:58:36Z"
    generation: 1
    name: k8sevents
    namespace: default
    resourceVersion: "363874"
    selfLink: /apis/sources.eventing.knative.dev/v1alpha1/namespaces/default/apiserversources/k8sevents
    uid: 0e0bd443-5965-47e5-8ee5-76382208f5dc
  spec:
    mode: Ref
    resources:
    - apiVersion: v1
      controller: false
      controllerSelector:
        apiVersion: ""
        kind: ""
        name: ""
        uid: ""
      kind: Event
      labelSelector: {}
    serviceAccountName: events-sa
    sink:
      ref:
        apiVersion: serving.knative.dev/v1
        kind: Service
        name: eshow1
        namespace: default
  status:
    conditions:
    - lastTransitionTime: "2020-02-25T16:59:20Z"
      status: "True"
      type: Deployed
    - lastTransitionTime: "2020-02-25T16:59:34Z"
      message: apiserversources.sources.eventing.knative.dev are deprecated and will
        be removed in the future. Use apiserversources.sources.knative.dev instead.
      reason: ApiServerSourceDeprecated
      severity: Warning
      status: "True"
      type: Deprecated
    - lastTransitionTime: "2020-02-25T16:58:53Z"
      severity: Info
      status: "True"
      type: EventTypesProvided
    - lastTransitionTime: "2020-02-25T16:59:20Z"
      status: "True"
      type: Ready
    - lastTransitionTime: "2020-02-25T16:58:53Z"
      status: "True"
      type: SinkProvided
    - lastTransitionTime: "2020-02-25T16:58:53Z"
      status: "True"
      type: SufficientPermissions
    observedGeneration: 1
    sinkUri: http://eshow1.default.svc.cluster.local
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

navidshaikh replied:
Following up our discussion on slack: Added a plain sources's common fields holding struct values we want to print for 'kn source list'. The human readable output stays same, but see
```
./kn source list -oyaml
Items:
- Name: k8sevents
  Namespace: default
  Ready: "True"
  Resource: apiserversources.sources.eventing.knative.dev
  Sink: svc:eshow1
  SourceKind: ApiServerSource
kind: List
```

navidshaikh commented on pkg/dynamic/client.go:
```diff
@@ -94,3 +111,86 @@ func (c *knDynamicClient) ListSourcesTypes() (*unstructured.UnstructuredList, er
+			return nil, err
+		}
+
+		sList, err := c.client.Resource(gvr).Namespace(namespace).List(options)
```
instantiated at the top once than for every element in loop

navidshaikh commented on pkg/dynamic/client.go:
```diff
@@ -94,3 +111,86 @@ func (c *knDynamicClient) ListSourcesTypes() (*unstructured.UnstructuredList, er
+func (c *knDynamicClient) ListSources(f *SourceListFilters) (*unstructured.UnstructuredList, error) {
+	var sourceList unstructured.UnstructuredList
+	options := metav1.ListOptions{}
+	sourceTypes, err := c.ListSourcesTypes()
```
All CRUD with dynamic client work with `Unstructured` objects, in this case, its kind is set to `CustomResourceDefinitionList` (see `kn source list-types -oyaml`).

rhuss replied:
I just wanted to say, for thus `UnstructuredList` we *know* the type (its `CustomeResourceDefinitionList` in advance, so could equally used a typed interface or marshall that list into the typed object.

For the source list its, different as it is a *heterogenous* list with different type of objects, so not really that fits well in the K8s model (a hint that this is not perfect is that you set the type of this arbitrarily to the type of the last item). The only thing they share is a common partial schema (duck type), so that fields from that partial schema can be printed out. Maybe you could introduced an object of this partial schema with the fields to printed out e.g `SourceDuck`, create a `SourceDuckList` type (which implements runtime.Object` that collect all the homogenous elements of the same object. 

Just a suggestion, we can iterate on this after this PR is merged (it gets a bit confusing otherwise).

navidshaikh commented on pkg/dynamic/client.go:
```diff
@@ -94,3 +111,86 @@ func (c *knDynamicClient) ListSourcesTypes() (*unstructured.UnstructuredList, er
+	return &sourceList, nil
+}
+
+func gvrFromUnstructured(u *unstructured.Unstructured) (gvr schema.GroupVersionResource, err error) {
```
There is `GroupVersionKind` for `Unstructured`, but didnt find one for GVR.
While writing this, I thought to write a conversion utility from GVK to GVR where 
`Resource` in GVR is plural of `Kind` in GVK.
We can replace `gvrFromUnstructured` with this ^^ conversion utility, WDYT ?

navidshaikh replied:
Well, we can't use GV from `GroupVersionKind()` received on this unstructured, as that would be of the CRD, not of the object which is defined in its spec.

navidshaikh commented on pkg/dynamic/client.go:
```diff
@@ -94,3 +111,86 @@ func (c *knDynamicClient) ListSourcesTypes() (*unstructured.UnstructuredList, er
+	if err != nil || !found {
+		return gvr, fmt.Errorf("can't find source GVR: %v", err)
+	}
+	version, found, err := unstructured.NestedString(content, "spec", "version")
```
following up on https://github.com/knative/client/pull/666#discussion_r382500719 comment, I think we should delegate finding GV to library instead and use conversion utility for Kind --> Resource.

Checked how `Version` is calculated here, its using `apiVersion` field and then parses group and version in it by splitting on `/`.

It doesn uses `version:` field, right! :+1:

navidshaikh commented on pkg/dynamic/client.go:
```diff
@@ -94,3 +111,86 @@ func (c *knDynamicClient) ListSourcesTypes() (*unstructured.UnstructuredList, er
+	if err != nil || !found {
+		return gvr, fmt.Errorf("can't find source GVR: %v", err)
+	}
+	resource, found, err := unstructured.NestedString(content, "spec", "names", "plural")
```
thats what I've observed so far..

rhuss commented on pkg/kn/commands/source/human_readable_flags.go:
```diff
@@ -85,3 +101,148 @@ func printSourceTypesList(sourceTypesList *unstructured.UnstructuredList, option
+}
+
+// printSourceList populates the source list table rows
+func printSourceList(sourceList *unstructured.UnstructuredList, options printers.PrintOptions) ([]metav1beta1.TableRow, error) {
```
So you could use a `sources []*Unstructured` as argument instead of `sourceList` here, too, as you only iterate of the list `items`, right ?

navidshaikh replied:
can't pass the slice as this function needs an object because:

https://github.com/knative/client/blob/a44833c849547a13048426b421b711267412db25/pkg/printers/interface.go#L26-L29

```
--- a/pkg/kn/commands/source/human_readable_flags.go
+++ b/pkg/kn/commands/source/human_readable_flags.go
@@ -123,14 +123,14 @@ func printSource(source unstructured.Unstructured, options printers.PrintOptions
 }
 
 // printSourceList populates the source list table rows
-func printSourceList(sourceList *unstructured.UnstructuredList, options printers.PrintOptions) ([]metav1beta1.TableRow, error) {
-	rows := make([]metav1beta1.TableRow, 0, len(sourceList.Items))
+func printSourceList(sourceList []unstructured.Unstructured, options printers.PrintOptions) ([]metav1beta1.TableRow, error) {
+	rows := make([]metav1beta1.TableRow, 0, len(sourceList))
 
-	sort.SliceStable(sourceList.Items, func(i, j int) bool {
-		return sourceList.Items[i].GetName() < sourceList.Items[j].GetName()
+	sort.SliceStable(sourceList, func(i, j int) bool {
+		return sourceList[i].GetName() < sourceList[j].GetName()
 	})
-	for _, item := range sourceList.Items {
-		row, err := printSource(item, options)
+	for _, source := range sourceList {
+		row, err := printSource(source, options)
 		if err != nil {
 			return nil, err
 		}
diff --git a/pkg/kn/commands/source/list.go b/pkg/kn/commands/source/list.go
index 2f157e17..c37dd0b3 100644
--- a/pkg/kn/commands/source/list.go
+++ b/pkg/kn/commands/source/list.go
@@ -65,7 +65,7 @@ func NewListCommand(p *commands.KnParams) *cobra.Command {
 			if err != nil {
 				return nil
 			}
-			err = printer.PrintObj(sourceList, cmd.OutOrStdout())
+			err = printer.PrintObj(sourceList.Items, cmd.OutOrStdout())
 			if err != nil {
 				return err
 			}
➜  client git:(pr/source-list) ✗ ./hack/build.sh -f
🚧 Compile
# knative.dev/client/pkg/kn/commands/source
pkg/kn/commands/source/list.go:68:37: cannot use sourceList.Items (type []unstructured.Unstructured) as type "k8s.io/apimachinery/pkg/runtime".Object in argument to printer.PrintObj:
	[]unstructured.Unstructured does not implement "k8s.io/apimachinery/pkg/runtime".Object (missing DeepCopyObject method)

```

rhuss replied:
Yes, I see. The funny thing is, that nothing really requires it to be a `runtime.Object` in the printed code. Its only in the print function which might extract this. So I wonder whether we just implement runtime.Object on the slice (with no-ops), like in

```golang
type sourceList []*Unstructured

func (s sourceList) GetObjectKind() {
    return EmptyObjectKind
}

func (s sourceList) DeepCopy(...) {
    ....
}
```
(you get the idea)

Regardless what, maybe just use the UnstructuredList for now, but set the the kind to `EmptyObjectKind`.

rhuss commented on pkg/kn/commands/source/human_readable_flags.go:
```diff
@@ -85,3 +101,148 @@ func printSourceTypesList(sourceTypesList *unstructured.UnstructuredList, option
+	return rows, nil
+}
+
+func findSink(source unstructured.Unstructured) string {
```
I think the `sink` should be always at the place in a source (as its a DuckType "Sinkable" probably ;-). So you could just navigate to those field into the unstructured source itself. But as mentioned elsewhere it might be better to move that kind of extraction allready when the list is queried from the server and introduce a view object that is just printed here. Putting too much of logic in the print functions makes it harder to maintain later.

rhuss commented on pkg/kn/commands/source/human_readable_flags.go:
```diff
@@ -85,3 +101,148 @@ func printSourceTypesList(sourceTypesList *unstructured.UnstructuredList, option
+	}
+}
+
+func isReady(source unstructured.Unstructured) string {
```
Same here. Whether the object is ready or not should be already extracted before the printing method is called.

rhuss commented on pkg/kn/commands/source/human_readable_flags.go:
```diff
@@ -85,3 +101,148 @@ func printSourceTypesList(sourceTypesList *unstructured.UnstructuredList, option
+}
+
+// temporary sinkToString for deprecated sources
+func sinkToString(sink *duckv1beta1.Destination) string {
```
I thought we already had an utility method for this ?

navidshaikh replied:
yes, but that util is for new sources using `duckv1.Destination`. I'd to write this for deprecated sources which are on `duckv1beta1.Destination`.

rhuss commented on pkg/kn/commands/source/human_readable_flags.go:
```diff
@@ -85,3 +101,148 @@ func printSourceTypesList(sourceTypesList *unstructured.UnstructuredList, option
+			return sink.URI.String()
+		}
+	}
+	return "<unknown>"
```
I would just would leave it empty. Its not "<unknown>" but just not configured.

navidshaikh commented:
- Updated PR to set empty apiVersion for list of different types of source objects for machine readable format (--type filter aware)
- Added `WithType` builder for dynamic client's `ListSources` to filter on source types.

See `kind` and `apiVersion` of `List` for machine readable format:

1. For multiple types of source objects
```
➜  client git:(pr/source-list) kn source list -oyaml| grep kind | tail -1
kind: List
```

```
➜  client git:(pr/source-list) kn source list -oyaml| grep apiVersion | head -1
apiVersion: ""
```

2. For single source type objects (or sub-set a single type using `--type` filter):
```
➜  client git:(pr/source-list) kn source list --type apiserversource -oyaml | grep kind | tail -1 
kind: ApiServerSourceList
```

```
➜  client git:(pr/source-list) kn source list --type apiserversource -oyaml | grep apiVersion | head -1
apiVersion: sources.eventing.knative.dev/v1alpha1
```

navidshaikh commented:
/retest

navidshaikh commented:
/test pull-knative-client-integration-tests

async revision delete issue

rhuss commented:
/retest

@navidshaikh there are some codegen errors, too

navidshaikh commented:
/retest

--- FAIL: TestSourceApiServer (210.14s)

rhuss commented:
@navidshaikh  lso codegen error. Guess that somehow sneaked in by a latest merge.

mattmoor commented on pkg/kn/commands/source/list_test.go:
```diff
@@ -0,0 +1,150 @@
+)
+
+// sourceFakeCmd takes cmd to be executed using dynamic client
+// pass the objects to be registed to dynamic clients
```
<!--typo-bot-->
Found potential misspelling(s):
```suggestion
// pass the objects to be registered to dynamic clients
```

rhuss commented on pkg/dynamic/client.go:
```diff
@@ -94,3 +99,57 @@ func (c *knDynamicClient) ListSourcesTypes() (*unstructured.UnstructuredList, er
+			sourceList.SetGroupVersionKind(sList.GetObjectKind().GroupVersionKind())
+		}
+	}
+	// Clear the Group and Version for list if there are multiple types of source objects found
```
👍

rhuss commented on pkg/dynamic/lib.go:
```diff
@@ -0,0 +1,125 @@
+)
+
+// gvrFromUnstructured takes a unstructured object of CRD type and finds GVR from its spec
+func gvrFromUnstructured(u *unstructured.Unstructured) (gvr schema.GroupVersionResource, err error) {
```
I'm still pretty sure that there must be a library call for doing this (I remember very darkly something like `GetGvrUnsafe()` or something like this. but  not sure).

I suggest to merge this in, but if we find out about that library replace it here.

rhuss replied:
my bad, its about parsing the spec's versions etc. which is kind of quite unique.

rhuss commented on test/e2e/source_apiserver_test.go:
```diff
@@ -55,6 +56,16 @@ func TestSourceApiServer(t *testing.T) {
+	t.Log("list sources in YAML format")
+	output = test.sourceList(t, r, "-oyaml")
+	assert.Check(t, util.ContainsAll(output, "testapisource1", "ApiServerSource", "Service", "testsvc0"))
+
```
thanks for already adding E2E tests, but maybe we could later enhance them with some more cases:

* Invalid filter option
* Heterogenous list of different types with `-o yaml` (would involve to combine a pingsource + apiserver source creation)
